plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/session-offset-high-water-mark.ts

```diff
@@ -0,0 +1,164 @@
+        const timeout = timeoutGuard(`${description} delayed. Waiting over 30 seconds.`)
+        try {
+            return await fn(client)
+        } catch (error) {
+            if (error instanceof SyntaxError) {
+                // invalid JSON
+                return null
+            } else {
+                throw error
+            }
+        } finally {
+            clearTimeout(timeout)
+            await this.redisPool.release(client)
+        }
+    }
+
+    public async add(tp: TopicPartition, sessionId: string, offset: number): Promise<void> {
+        const key = offsetHighWaterMarkKey(this.keyPrefix, tp)
+        try {
+            await this.run(`write offset high-water mark ${key} `, async (client) => {
+                const returnCountOfUpdatedAndAddedElements = 'CH'
+                const updatedCount = await client.zadd(key, returnCountOfUpdatedAndAddedElements, offset, sessionId)
+                status.info('📝', 'WrittenOffsetCache added high-water mark for partition', {
+                    key,
+                    ...tp,
+                    sessionId,
+                    offset,
+                    updatedCount,
+                })
+                this.topicPartitionWaterMarks[`${tp.topic}-${tp.partition}`][sessionId] = offset
+            })
+        } catch (error) {
+            status.error('🧨', 'WrittenOffsetCache failed to add high-water mark for partition', {
+                error: error.message,
+                key,
+                ...tp,
+                sessionId,
+                offset,
+            })
+            captureException(error, {
+                extra: {
+                    key,
+                    offset,
+                },
+                tags: {
+                    ...tp,
+                    sessionId,
+                },
+            })
+        }
+    }
+
+    public async getAll(tp: TopicPartition): Promise<Record<string, number> | null> {
+        const key = offsetHighWaterMarkKey(this.keyPrefix, tp)
+        try {
+            return await this.run(`read all offset high-water mark for ${key} `, async (client) => {
+                const redisValue = await client.zrange(key, 0, -1, 'WITHSCORES')
+                // redisValue is an array of [sessionId, offset, sessionId, offset, ...]
+                // we want to convert it to an object of { sessionId: offset, sessionId: offset, ... }
+                const highWaterMarks = redisValue.reduce(
+                    (acc: Record<string, number>, value: string, index: number) => {
+                        if (index % 2 === 0) {
+                            acc[value] = parseInt(redisValue[index + 1])
+                        }
+                        return acc
+                    },
+                    {}
+                )
+                if (highWaterMarks) {
+                    this.topicPartitionWaterMarks[`${tp.topic}-${tp.partition}`] = highWaterMarks
+                }
```

Comment:
I don't think the if makes sense. We should _always_ set a value otherwise we can end up stuck with every single message calling redis